File: 415.字符串相加.rs
/*
 * @lc app=leetcode.cn id=415 lang=rust
 *
 * [415] 字符串相加
 */

// @lc code=start
impl Solution {
    pub fn add_strings(num1: String, num2: String) -> String {
        let (num1, num2) = (
            num1.chars().collect::<Vec<char>>(),
            num2.chars().collect::<Vec<char>>(),
        );
        let mut ret = String::new();
        let (mut carry, mut i, mut j) = (0, num1.len() as i32 - 1, num2.len() as i32 - 1);

        while i >= 0 || j >= 0 || carry > 0 {
            let digit1 = if i >= 0 {
                num1[i as usize] as u8 - b'0'
            } else {
                0
            };
            let digit2 = if j >= 0 {
                num2[j as usize] as u8 - b'0'
            } else {
                0
            };

            let sum = digit1 + digit2 + carry;
            carry = sum / 10;

            ret.insert(0, (sum % 10 + b'0') as char);

            i -= 1;
            j -= 1;
        }
        ret
    }
}
// @lc code=end

File: leetcode/415.字符串相加.ts
/*
 * @lc app=leetcode.cn id=415 lang=typescript
 *
 * [415] 字符串相加
 */

// @lc code=start
function addStrings(num1: string, num2: string): string {
    let [i1, i2, carry] = [num1.length-1, num2.length-1, 0];
    let ret = "";
    while (i1 >= 0 || i2 >= 0 || carry > 0) {
        let [n1, n2] = [0, 0];
        if (i1 >= 0) n1 = num1.charCodeAt(i1) - 48;
        if (i2 >= 0) n2 = num2.charCodeAt(i2) - 48;

        const sum = n1 + n2 + carry;
        ret = sum%10 + ret;
        carry = Math.trunc(sum/10);
        --i1;
        --i2;
    }
    return ret;
};
// @lc code=end

